### PR TITLE
Refactor PackageDependencies property

### PR DESCRIPTION
**Pull request by scottmitchell**
### Purpose
This PR addresses https://jira.autodesk.com/browse/DYN-1997 and also simplifies the package dependency collection process. 
This PR removes `LoadedPackageDependencies` property and refactors `PackageDependencies` property/collection and `PackageDependencyInfo` object.

- Only one package should be listed as a dependency for a node
- By default, the original deserialized package dependency should be maintained
- API to override the original dependency with the local dependency
- Unit test
- Adds new `PackageInfo` type.
- Adds `IsLoaded` property to `PackageDependencyInfo`
- Simplifies package dependency collection process

This should fix the case found by @QilongTang where a node can have two package dependencies when multiple packages are loaded with the same custom node. Previously, both the deserialized package dependency and the collected package dependency of a single node could be contained in the `PackageDependencies` property of a workspace. Now, the *original* package dependency is maintained until it is voided using `VoidNodeDependency()`. After the original dependency has been voided, the local package dependency will be used for serialization.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @aparajit-pratap @QilongTang @ColinDayOrg

### CODE REVIEW

**Comment by mjkkirschner** [comment]
so this does not address the underlying issue right? (as you said above) - the point brought up by @ColinDayOrg - that likely which package is loaded depends on load order. IE if you restart Dynamo you might load packages in a different order than if install them manually.

File: src/DynamoCore/Graph/Workspaces/PackageDependencyInfo.cs [hunk 1]
```diff
@@ -3,21 +3,97 @@
 
 namespace Dynamo.Graph.Workspaces
 {
+    internal class PackageInfo
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/PackageDependencyInfo.cs]: even though it's internal, a summary block would not hurt.

**Comment by scottmitchell** [comment]
@mjkkirschner Right, this PR just allows package dependency collection to deal with cases where multiple packages, or versions of packages, are able to resolve nodes. The underlying issue will be addressed later, as the outcome of @aparajit-pratap 's spike into the issue.

**Comment by mjkkirschner** [comment]
@scottmitchell this looks good to me - can you also cherry pick to 2.3 branch.

File: src/PackageDependencyViewExtension/PackageDependencyView.xaml.cs [hunk 2]
```diff
@@ -65,26 +65,18 @@ internal void DependencyRegen(WorkspaceModel ws)
             table.Columns.Clear();
             foreach (var package in ws.PackageDependencies)
             {
-                bool matchFound = false;
-                // Check if the target package is installed and loaded.
-                foreach (var loadedPackage in ws.LoadedPackageDependencies)
+                if (package.IsLoaded)
```
Comment (QilongTang) [inline src/PackageDependencyViewExtension/PackageDependencyView.xaml.cs]: Much easier to refresh UI now, great. The only question is that how costly to maintain this property on each node add operation, would love to know how slow it is

**Comment by QilongTang** [comment]
LGTM

**Comment by aparajit-pratap** [comment]
@scottmitchell I have a DYN file with package dependency as clockwork 1.x and the same package installed. However when I load the file, I still see package dependency 1.x to be missing (in red). I tried the same thing with 2.x; existing DYN file with package dependency serialized as 2.x, and 2.x locally installed. Same issue; 2.x is still missing in view extension.

**Comment by mjkkirschner** [comment]
@scottmitchell since we all missed that bug except @aparajit-pratap 🦅 - can we add a test for it?

**Pull request merged**